src/vocabulary.rs

```diff
@@ -0,0 +1,519 @@
+    /// They have been installed.
+    InstalledMissing,
+
+    /// The vocabulary was present, at the correct version, and all attributes were present.
+    Existed,
+
+    /// The vocabulary was present, at an older version, and it has been upgraded. Any
+    /// missing attributes were installed.
+    Upgraded,
+}
+
+pub trait HasVocabularies {
+    fn read_vocabularies(&self) -> Result<Vocabularies>;
+    fn read_vocabulary_named(&self, name: &NamespacedKeyword) -> Result<Option<Vocabulary>>;
+}
+
+pub trait VersionedStore {
+    /// Check whether the vocabulary described by the provided metadata is present in the store.
+    fn check_vocabulary<'definition>(&self, definition: &'definition Definition) -> Result<VocabularyCheck<'definition>>;
+
+    /// Check whether the provided vocabulary is present in the store. If it isn't, make it so.
+    fn ensure_vocabulary(&mut self, definition: &Definition) -> Result<VocabularyOutcome>;
+
+    /// Make sure that our expectations of the core vocabulary -- basic types and attributes -- are met.
+    fn verify_core_schema(&self) -> Result<()>;
+}
+
+trait VocabularyMechanics {
+    fn install_vocabulary(&mut self, definition: &Definition) -> Result<VocabularyOutcome>;
+    fn install_attributes_for<'definition>(&mut self, definition: &'definition Definition, attributes: Vec<&'definition (NamespacedKeyword, Attribute)>) -> Result<VocabularyOutcome>;
+    fn upgrade_vocabulary(&mut self, definition: &Definition, from_version: Vocabulary) -> Result<VocabularyOutcome>;
+}
+
+impl Vocabulary {
+    // TODO: don't do linear search!
+    fn find<T>(&self, entid: T) -> Option<&Attribute> where T: Into<Entid> {
+        let to_find = entid.into();
+        self.attributes.iter().find(|&&(e, _)| e == to_find).map(|&(_, ref a)| a)
+    }
+}
+
+impl<'a, 'c> VersionedStore for InProgress<'a, 'c> {
+    fn verify_core_schema(&self) -> Result<()> {
+        if let Some(core) = self.read_vocabulary_named(&DB_SCHEMA_CORE)? {
+            if core.version != CORE_SCHEMA_VERSION {
+                bail!(ErrorKind::UnexpectedCoreSchema(Some(core.version)));
+            }
+
+            // TODO: check things other than the version.
+        } else {
+            // This would be seriously messed up.
+            bail!(ErrorKind::UnexpectedCoreSchema(None));
+        }
+        Ok(())
+    }
+
+    fn check_vocabulary<'definition>(&self, definition: &'definition Definition) -> Result<VocabularyCheck<'definition>> {
+        if let Some(vocabulary) = self.read_vocabulary_named(&definition.name)? {
+            // The name is present.
+            // Check the version.
+            if vocabulary.version == definition.version {
+                // Same version. Check that all of our attributes are present.
+                let mut missing: Vec<&'definition (NamespacedKeyword, Attribute)> = vec![];
+                for pair in definition.attributes.iter() {
+                    if let Some(entid) = self.get_entid(&pair.0) {
```

Comment:
When does this return `None`?  (Never, right?)  This loop reads a little odd, since the default case looks like it happens too frequently.  Perhaps we should: lift the default into an else for the `Some(existing)` for clarity, and bail if we can't bind `self.get_entid(...)`, since something bad is happening if that's the case.

Reply:
It will return `None` when an attribute is added to a vocabulary — the attribute's ident will be missing in the store.

You might be mixing up `definition.name`, which is the name of the vocabulary (e.g., `org.mozilla.bookmarks`) and `pair.0`, which is the ident of an attribute (e.g., `:foo/title`).